### PR TITLE
Livstream: support for downloading videos with account id

### LINKED ISSUE #2225: [livestream extractor] Unable to download JSON metadata (HTTP Error 404)
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2021.12.27**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Description

Downloading of https://livestream.com/accounts/[user-number] gives me an HTTP Error 404. I'd like to download several videos from the account. Individual videos are working.

### Verbose log

```shell
yt-dlp -U --verbose  https://livestream.com/accounts/8712572
[debug] Command-line config: ['-U', '-o', 'C:/Users/XYZ/Downloads/test_DL/output/%(upload_date>%y%m%d)s_%(title)s.%(ext)s', '--verbose', 'https://livestream.com/accounts/8712572']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2021.12.27 [6223f67] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19043-SP0
[debug] exe versions: ffmpeg 2021-08-10-git-c245963a21-full_build-www.gyan.dev (setts), ffprobe 2021-08-10-git-c245963a21-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2021.12.27, Current version: 2021.12.27
yt-dlp is up to date (2021.12.27)
[debug] [livestream] Extracting URL: https://livestream.com/accounts/8712572
[livestream] Downloading JSON metadata
ERROR: [livestream] Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: 'Not Found'>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; type  yt-dlp -U  to update. Be sure to call yt-dlp with the --verbose flag and include its complete output. (caused by <HTTPError 404: 'Not Found'>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; type  yt-dlp -U  to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.
  File "yt_dlp\extractor\common.py", line 717, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3458, in urlopen
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 563, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 755, in http_error_302
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default


C:\Users\Puresive Films>
```

### PR DESCRIPTION
**Pull request by theperfectpunk**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Fixes livestream test cases and adds support for downloading videos of events of an account

Fixes #2225

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/livestream.py [hunk 1]
```diff
@@ -22,12 +22,12 @@
 
 class LivestreamIE(InfoExtractor):
     IE_NAME = 'livestream'
-    _VALID_URL = r'https?://(?:new\.)?livestream\.com/(?:accounts/(?P<account_id>\d+)|(?P<account_name>[^/]+))/(?:events/(?P<event_id>\d+)|(?P<event_name>[^/]+))(?:/videos/(?P<id>\d+))?'
+    _VALID_URL = r'https?://(?:new\.)?livestream.com/(?:(?:accounts/(?P<account_id>\d+))|(?P<account_name>[^/]+))(?:(?:/events/(?P<event_id>\d+))|(?:/(?P<event_name>[^/]+)))?(?:/videos/(?P<id>\d+))?'
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py:25]: Hard to read. could u split it up into multiple lines?

File: yt_dlp/extractor/livestream.py [hunk 2]
```diff
@@ -207,21 +223,39 @@ def _extract_event(self, event_data):
...
+        api_url = self._ACCOUNT_API_URL_TEMPLATE % (account)
         if video_id:
             video_data = self._download_json(
-                api_url + '/videos/%s' % video_id, video_id)
+                api_url + '/events/%s' % event + '/videos/%s' % video_id, video_id)
             return self._extract_video_info(video_data)
-        else:
-            event_data = self._download_json(api_url, video_id)
+        if event:
+            event_data = self._download_json(api_url + '/events/%s' % event, None)
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py:255]:
```suggestion
        api_url = f'http://livestream.com/api/accounts/{account}'
        if video_id:
            video_data = self._download_json(
                 f'{api_url}/events/{event}/videos/{video_id}', video_id)
            return self._extract_video_info(video_data)
        elif event:
            event_data = self._download_json(f'{api_url}/events/{event}', None)
```

File: yt_dlp/extractor/livestream.py [hunk 3]
```diff
@@ -207,21 +223,39 @@ def _extract_event(self, event_data):
...
+            playlist = self._generate_event_playlist(event_data)
+            if playlist:
+                entries = entries + playlist
+        return self.playlist_result(entries, account_data['id'], account_data['full_name'])
+
     def _real_extract(self, url):
         mobj = self._match_valid_url(url)
         video_id = mobj.group('id')
         event = mobj.group('event_id') or mobj.group('event_name')
         account = mobj.group('account_id') or mobj.group('account_name')
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py:248]: This needs a rewrite, but I won't insuiste coz it's legacy code

File: yt_dlp/extractor/livestream.py [hunk 4]
```diff
@@ -207,21 +223,39 @@ def _extract_event(self, event_data):
                     'http://livestream.com/accounts/%s/events/%s/videos/%s' % (account_id, event_id, v_id),
                     'Livestream', v_id, v.get('caption')))
             last_video = videos_info[-1]['id']
+        return entries
+
+    def _extract_event(self, event_data):
+        event_id = compat_str(event_data['id'])
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py:229]: compat_str is deprecated

File: yt_dlp/extractor/livestream.py [hunk 5]
```diff
@@ -207,21 +223,39 @@ def _extract_event(self, event_data):
...
+    def _extract_account_events(self, account_data):
+        upcoming_events = account_data['upcoming_events']
+        past_events = account_data['past_events']
+        entries = []
+        events = upcoming_events['data'] + past_events['data']
+        for event_data in events:
+            playlist = self._generate_event_playlist(event_data)
+            if playlist:
+                entries = entries + playlist
+        return self.playlist_result(entries, account_data['id'], account_data['full_name'])
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py:242]:
```suggestion
    def _extract_account_events(self, account_data):
        items = traverse_obj(account_data, (('upcoming_events', 'past_events'), 'data', ...))
        return self.playlist_result(
            itertools.chain.from_iterable(map(self._generate_event_playlist, items),
            account_data.get('id'), account_data.get('full_name'))
```
Don't make this a function either

**Review by pukkandan — CHANGES_REQUESTED** [review]
I've directly made code suggestions. But next time, pls **actually** read the contributing guidelines

File: yt_dlp/extractor/livestream.py (hunk 1 above)
Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py]: done

File: yt_dlp/extractor/livestream.py (hunk 2 above)
Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py]: added suggested code

File: yt_dlp/extractor/livestream.py (hunk 4 above)
Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py]: replaced with str

File: yt_dlp/extractor/livestream.py [hunk 6]
```diff
@@ -17,17 +17,21 @@
...
 )
 
 
 class LivestreamIE(InfoExtractor):
     IE_NAME = 'livestream'
-    _VALID_URL = r'https?://(?:new\.)?livestream\.com/(?:accounts/(?P<account_id>\d+)|(?P<account_name>[^/]+))/(?:events/(?P<event_id>\d+)|(?P<event_name>[^/]+))(?:/videos/(?P<id>\d+))?'
+    _VALID_URL = r'https?://(?:new\.)?livestream.com/' \
+        '(?:(?:accounts/(?P<account_id>\d+))|(?P<account_name>[^/]+))' \
+        '(?:(?:/events/(?P<event_id>\d+))|(?:/(?P<event_name>[^/]+)))?' \
+        '(?:/videos/(?P<id>\d+))?'
```
Comment (pukkandan) [inline yt_dlp/extractor/livestream.py]:
```suggestion
    _VALID_URL = r'''(?x)
        https?://(?:new\.)?livestream.com/
        (?:accounts/(?P<account_id>\d+)|(?P<account_name>[^/]+))
        (?:/events/(?P<event_id>\d+)|/(?P<event_name>[^/]+))?
        (?:/videos/(?P<id>\d+))?
    '''
```

Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py]: changes done

File: yt_dlp/extractor/livestream.py (hunk 3 above)
Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py:248]: @pukkandan I'll do the rewrite, please point me to a reference elsewhere in the code. I'll follow the same pattern.

File: yt_dlp/extractor/livestream.py (hunk 5 above)
Comment (theperfectpunk) [reply yt_dlp/extractor/livestream.py]: @pukkandan I can get rid of the functions and move the logic to _real_extract, to make it look like

```python3
def _real_extract(self, url):
    mobj = self._match_valid_url(url)
    video_id = mobj.group('id')
    event = mobj.group('event_id') or mobj.group('event_name')
    account = mobj.group('account_id') or mobj.group('account_name')
    api_url = f'http://livestream.com/api/accounts/{account}'
    if video_id:
        video_data = self._download_json(
            f'{api_url}/events/{event}/videos/{video_id}', video_id)
        return self._extract_video_info(video_data)
    elif event:
        event_data = self._download_json(f'{api_url}/events/{event}', None)
        event_id = str(event_data['id'])
        entries = self._generate_event_playlist(event_data)
        return self.playlist_result(entries, event_id, event_data['full_name'])
    account_data = self._download_json(api_url, None)
    items = traverse_obj(account_data, (('upcoming_events', 'past_events'), 'data', ...))
    return self.playlist_result(
        itertools.chain.from_iterable(map(self._generate_event_playlist, items)),
        account_data.get('id'), account_data.get('full_name'))
```

**Pull request merged**